allennlp/models/multitask.py

```diff
@@ -84,58 +84,59 @@ def __init__(
         super().__init__(vocab, **kwargs)
         self._backbone = backbone
         self._heads = torch.nn.ModuleDict(heads)
+        self._heads_called: Set[str] = set()
         self._arg_name_mapping = arg_name_mapping or defaultdict(dict)
 
         self._allowed_arguments = allowed_arguments or {
             "backbone": get_forward_arguments(backbone),
             **{key: get_forward_arguments(heads[key]) for key in heads},
         }
         self._loss_weights = loss_weights or defaultdict(lambda: 1.0)
-        self._active_heads: Optional[List[str]] = None
         initializer(self)
 
-    def set_active_heads(self, active_heads: List[str]) -> None:
-        """
-        By default, the `MultiTaskModel` will try to infer which heads to run from the arguments
-        passed to `forward`.  During training, we will only run a head if we have all of its
-        arguments, including optional arguments, which typically means the argument is the
-        prediction target; if we don't have it, we typically can't compute a loss, so running during
-        training is pointless.  During evaluation, we will run all heads.
-
-        If you want to limit which heads are run during evaluation, or if the inference for which
-        task to run during training is incorrect (e.g., if your head has multiple optional
-        arguments, and only some are actually required to compute a loss), then you can use this
-        method to override our inference and force the use of whatever set of heads you want.
-
-        To get back to the default mode of operation, call this method with `None` as an argument.
-        """
-        self._active_heads = active_heads
-
     def forward(self, **kwargs) -> Dict[str, torch.Tensor]:  # type: ignore
+        if "task" not in kwargs:
+            raise ValueError(
+                "Instances for multitask training need to contain a MetadataField with "
+                "the name 'task' to indicate which task they belong to. Usually the "
+                "MultitaskDataLoader provides this field and you don't have to do anything."
+            )
+
+        task_indices_just_for_mypy: Mapping[str, List[int]] = defaultdict(lambda: [])
+        for i, task in enumerate(kwargs["task"]):
+            task_indices_just_for_mypy[task].append(i)
+        task_indices: Dict[str, torch.LongTensor] = {
+            task: torch.LongTensor(indices) for task, indices in task_indices_just_for_mypy.items()
+        }
+
+        def make_inputs_for_task(task: str, whole_batch_input: Union[torch.Tensor, List]):
+            if isinstance(whole_batch_input, torch.Tensor):
+                task_indices[task] = task_indices[task].to(whole_batch_input.device)
+                return torch.index_select(whole_batch_input, 0, task_indices[task])
+            else:
+                return [whole_batch_input[i] for i in task_indices[task]]
```

Comment:
This machinery splits up the batch by head, and runs the heads separately. I wonder if in this way we can support heterogeneous batches more or less transparently and for all models, though in the interest of time I did not try this.